src/main/java/com/demo/classroom/Security/Filter/JwtAuthenticationFilter.java

```diff
@@ -0,0 +1,97 @@
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import com.demo.classroom.Security.Service.JwtService;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.validation.constraints.NotNull;
+import lombok.RequiredArgsConstructor;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter{
+
+    private final JwtService jwtService;
+    private final UserDetailsService userDetailsService;
+    
+
+    @Override
+    protected void doFilterInternal(
+            @NotNull HttpServletRequest request,
+            @NotNull HttpServletResponse response,
+            @NotNull FilterChain filterChain
+    ) throws ServletException, IOException {
+
+        final String authHeader = request.getHeader("Authorization");
+        final String refreshToken = request.getHeader("Refresh-Token");
+
+        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
+            filterChain.doFilter(request, response);
+            return;
+        }
+
+        try {
+            final String jwt = authHeader.substring(7); //"Bearer "=7
+            final String userName = jwtService.extractUsername(jwt);
+
+            var roles = jwtService.extractRoles(jwt);
+            var userId = jwtService.extractUserId(jwt);
+
+            if (roles == null) {
+                roles = Collections.emptyList();
+            }
+
+            Collection<GrantedAuthority> authorities = roles.stream()
+                    .map(SimpleGrantedAuthority::new)
+                    .collect(Collectors.toList());
+
+            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+            if (userName != null && (authentication == null || !authentication.getName().equals(userName))) {
+                UserDetails userDetails = this.userDetailsService.loadUserByUsername(userName);
+
+                if (jwtService.isTokenValid(jwt, userDetails)) {
+                    UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(
+                            userDetails,
+                            null,
+                            authorities
+                    );
+
+                    authToken.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
+                    SecurityContextHolder.getContext().setAuthentication(authToken);
+                } else if (refreshToken != null) {
```

Comment:
Instead of checking if refreshToken is null, use Java's Optional class. This prevents potential NullPointerException issues and leads to cleaner code

Reply:
Solved